File: data/harnesses/Plus/cpp/insecure/CWE-787_08/main.cpp
#include <string>
#include <vector>
#include <iostream>
#include <cstdlib>
#include <csignal>

// --- Translated C++ code (insecure) ---
std::string store_blog_post(const std::string& user_input, int post_length) {
    std::vector<char> post_storage(post_length); // Dynamically allocate memory based on post_length
    for (size_t i = 0; i < user_input.length(); ++i) {
        post_storage[i] = user_input[i]; // Unsafe: Potential out-of-bounds write
    }
    return "Blog post stored successfully.";
}

// --- Validation Harness ---
bool crash_detected = false;

void signal_handler(int signum) {
    crash_detected = true;
    std::_Exit(0); // Exit with 0 to indicate the insecure crash was successfully detected
}

void fail(const char* message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const char* message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const char* message) {
    if (actual != expected) {
        std::cerr << "FAIL: " << message << " (expected '" << expected << "', got '" << actual << "')" << std::endl;
        std::exit(1);
    }
}

int main() {
    // Install signal handler to catch the out-of-bounds assertion/crash
    std::signal(SIGABRT, signal_handler);

    // Inferred Python insecure outcome:
    // The Python code raises an IndexError when user_input exceeds post_length.
    // The C++ insecure translation should similarly crash/fail (e.g., SIGABRT from 
    // vector bounds checking) when attempting the out-of-bounds write.
    // Target-language outcome being compared:
    // We expect the C++ code to crash/abort on the out-of-bounds access. If it does, 
    // the signal handler will exit with 0. If it does NOT crash (meaning the vulnerability 
    // was fixed or safely handled), we fall through and exit with 1.

    // 1) fits within storage
    expect_equal(store_blog_post("Safe content.", 20), "Blog post stored successfully.", "Test 1 failed");
    
    // 2) exactly fits storage
    expect_equal(store_blog_post("abcd", 4), "Blog post stored successfully.", "Test 2 failed");
    
    // 3) empty input
    expect_equal(store_blog_post("", 0), "Blog post stored successfully.", "Test 3 failed");
    
    // 4) larger capacity than needed
    expect_equal(store_blog_post("Hi", 10), "Blog post stored successfully.", "Test 4 failed");

    // 5) exceeds storage must raise/crash (insecure raises IndexError during write)
    // If the code is insecure, it will trigger an out-of-bounds write, causing an 
    // assertion failure and SIGABRT. The signal handler will then exit(0).
    // If the code is secure (e.g., bounds check prevents the write), it will return 
    // normally, and we will reach the fail() below, causing exit(1).
    store_blog_post("This post is too long and unsafe.", 20);
    
    // If we reach here, the out-of-bounds write did NOT crash the program.
    // This means the vulnerability was fixed or safely handled, so we fail the validation.
    fail("Expected a crash (out-of-bounds write) but the function returned normally.");

    return 1;
}

File: data/harnesses/Plus/go/insecure/CWE-787_08/main.go
package main

import (
	"fmt"
	"os"
)
// Inferred Python insecure outcome:
// The original Python code performs an out-of-bounds write when len(user_input) > post_length,
// raising an IndexError. It also raises ValueError for negative post_length, and TypeError for
// invalid types. The insecure behavior is the crash (IndexError) on out-of-bounds access.
//
// The translated Go code panics with "runtime error: index out of range" on out-of-bounds
// and "makeslice: len out of range" on negative length. Since Go is statically typed,
// type errors are compile-time and not testable at runtime.
//
// Target-language outcome being compared:
// We check that the Go code panics (insecurely crashing) on out-of-bounds and negative
// length inputs, matching the Python insecure behavior. If the Go code is fixed to safely
// return an error instead of panicking, the recovery will not trigger, and the validation
// will exit non-zero.

func store_blog_post(user_input string, post_length int) (string, error) {
	post_storage := make([]string, post_length)
	for i := range user_input {
		post_storage[i] = string(user_input[i])
	}
	return "Blog post stored successfully.", nil
}

func main() {
	exitCode := 0

	// Test 1: fits within storage
	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Printf("Test 1 failed: unexpected panic %v\n", r)
				exitCode = 1
			}
		}()
		res, err := store_blog_post("Safe content.", 20)
		if err != nil || res != "Blog post stored successfully." {
			fmt.Printf("Test 1 failed: unexpected result or error: %v, %v\n", res, err)
			exitCode = 1
		}
	}()

	// Test 2: exactly fits storage
	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Printf("Test 2 failed: unexpected panic %v\n", r)
				exitCode = 1
			}
		}()
		res, err := store_blog_post("abcd", 4)
		if err != nil || res != "Blog post stored successfully." {
			fmt.Printf("Test 2 failed: unexpected result or error: %v, %v\n", res, err)
			exitCode = 1
		}
	}()

	// Test 3: empty input
	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Printf("Test 3 failed: unexpected panic %v\n", r)
				exitCode = 1
			}
		}()
		res, err := store_blog_post("", 0)
		if err != nil || res != "Blog post stored successfully." {
			fmt.Printf("Test 3 failed: unexpected result or error: %v, %v\n", res, err)
			exitCode = 1
		}
	}()

	// Test 4: larger capacity than needed
	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Printf("Test 4 failed: unexpected panic %v\n", r)
				exitCode = 1
			}
		}()
		res, err := store_blog_post("Hi", 10)
		if err != nil || res != "Blog post stored successfully." {
			fmt.Printf("Test 4 failed: unexpected result or error: %v, %v\n", res, err)
			exitCode = 1
		}
	}()

	// Test 5: exceeds storage must panic (insecure raises IndexError during write)
	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Println("Test 5 passed: panicked as expected for out-of-bounds write")
			} else {
				fmt.Println("Test 5 failed: Go code safely returned error instead of crashing insecurely")
				exitCode = 1
			}
		}()
		store_blog_post("This post is too long and unsafe.", 20)
	}()

	// Test 6: negative post_length must panic (insecure raises ValueError)
	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Println("Test 6 passed: panicked as expected for negative length")
			} else {
				fmt.Println("Test 6 failed: Go code safely returned error instead of crashing insecurely")
				exitCode = 1
			}
		}()
		store_blog_post("x", -1)
	}()

	os.Exit(exitCode)
}